server/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregator.java

```diff
@@ -113,22 +114,54 @@ protected PipelineAggregator(String name, String[] bucketsPaths, Map<String, Obj
 
     /**
      * Read from a stream.
+     * @deprecated pipeline aggregations added after 7.8.0 shouldn't call this
      */
+    @Deprecated
     protected PipelineAggregator(StreamInput in) throws IOException {
-        name = in.readString();
-        bucketsPaths = in.readStringArray();
-        metadata = in.readMap();
+        if (in.getVersion().before(Version.V_7_8_0)) {
+            name = in.readString();
+            bucketsPaths = in.readStringArray();
+            metadata = in.readMap();
+        } else {
+           throw new IllegalArgumentException("[" + getClass() + "] is not supported on versions before 7.8.0");
```

Comment:
Hmm, I think these exception messages are a bit confusing.  If I understand correctly, this path theoretically should never happen because e.g. InternalAggregations won't serialize pipelines in 7.8+, so read/write on the aggregator 7.8+ should never be invoked.

I think we should probably change the message to "Cannot (de)serialize pipeline aggregator from stream because pipelines are not serialized after 7.8" or something to that effect.  That way if a user does see this, it won't confuse them; the current message may make them think the aggregation itself is no longer supported (due to the node being 7.8+ to reach that statement, but the message saying "before 7.8").

I think we need to throw a 5xx error too, because if we get here we are firmly in the server-side going pear shaped and not the fault of a client/user.  Perhaps `IllegalStateException`?

Reply:
Makes sense to me!